File: examples/languages/go/order_service.go
package orders

type Product struct {
	Price int
}

type Order struct {
	ProductID string
	Quantity  int
	Total     int
}

type Catalog interface {
	Find(productID string) *Product
}

type OrderRepository interface {
	Save(order Order)
}

type OrderService struct {
	catalog    Catalog
	repository OrderRepository
}

func (service *OrderService) Create(productID string, quantity int) Order {
	product := service.catalog.Find(productID)
	if product == nil {
		panic("unknown product")
	}

	order := Order{
		ProductID: productID,
		Quantity:  quantity,
		Total:     product.Price * quantity,
	}
	service.repository.Save(order)
	return order
}

File: examples/languages/rust/order_service.rs
pub struct Product {
    pub price: u64,
}

pub struct Order {
    pub product_id: String,
    pub quantity: u32,
    pub total: u64,
}

pub trait Catalog {
    fn find(&self, product_id: &str) -> Option<Product>;
}

pub trait OrderRepository {
    fn save(&self, order: &Order);
}

pub struct OrderService {
    catalog: Box<dyn Catalog>,
    repository: Box<dyn OrderRepository>,
}

impl OrderService {
    pub fn create(&self, product_id: &str, quantity: u32) -> Order {
        let product = self.catalog.find(product_id);
        if product.is_none() {
            panic!("unknown product");
        }

        let product = product.unwrap();
        let order = Order {
            product_id: product_id.to_owned(),
            quantity,
            total: product.price * quantity as u64,
        };
        self.repository.save(&order);
        return order;
    }
}
